ebean-core/src/main/java/io/ebeaninternal/server/transaction/DefaultPersistenceContext.java

```diff
@@ -279,6 +280,10 @@ private void put(Object id, Object bean) {
       Object existing;
       if (useReferences) {
         weakCount++;
+        if (id instanceof EntityBean) {
+          // we are an embedded ID
+          id = cloneEmbeddedId((EntityBean) id);
+        }
         existing = map.put(id, new BeanRef(this, id, bean, queue));
```

Comment:
It still can run out of memory if ids are quite large

Reply:
Unfortunately I can't quite follow you here. Do you mean "one ID is big" or do you mean "many IDs are in the cache"?

- "one ID is big": An ID should not be big in DB terms. You should keep your primary keys small. There are often limitations: MariaDB can only use 16 columns and has a max key length of 767 bytes. So I think an ID itself should always fit in < 1kb of ram. For me it makes also no sense to have large text fields as primary key.
- So I'm sure you ment "many IDs": There should not be too much entries in the cache. If the memory goes low, the gc will put the `BeanRefs` in the refQueue (if there are no strong-references of course) and `DefaultPersistenceContext.expungeStaleEntries` will remove the whole map entry.

If I've missed something here, please report back :)

Nevertheless, this PR is not the best solution, because there are many caches in ebean (and in our app) that have the same problem if embedded IDs were used.

cheers
Roland

Reply:
@rPraml I mean, `map` is a regular HashMap. Id is put there as a "strong" reference. As I understand context is not cleared until the streaming is over. So even if we have "weak" values in map, the key set itself may cause OutOfMemory. It will just take longer.

Reply:
> the key set itself may cause OutOfMemory

No. As Jonas points out the key is removed.

Reply:
@rbygrave Why do you need to cache entities while streaming in the first place?

Reply:
You need it, to guarantee that you get the same object.

Assume you iterate over invoices and they reference a customer.
So you can have two invoices pointing to the same customer.
What we want is, that we don't have multiple instances of the same customer in memory. They should all point to the same object.

You are partially right, that we do not need this in findeach... as long as you do not hold references to previous objects.
Assume that you try to compare previous object with the current one, while iterating.

Reply:
@rPraml Interesting. I never expect it to be the same object, to be honest. Each entity should have proper `equals()` and `hashCode()` implementations. That's it.

Reply:
If the projection includes a ManyToOne, OneToOne, OneToMany or ManyToMany we need the persistence context to de-duplicate the result in order to produce consistent object graphs.

This is the fundamental difference between an ORM query and a 'row mapping' query like DtoQuery.

If your query does not include any relationships then yes the persistence context isn't doing any de-duplication.